File: samplecode/mono/Naive.cs
//    Copyright 2014 athenahealth, Inc.
//
//   Licensed under the Apache License, Version 2.0 (the "License"); you
//   may not use this file except in compliance with the License.  You
//   may obtain a copy of the License at
//
//       http://www.apache.org/licenses/LICENSE-2.0
//
//   Unless required by applicable law or agreed to in writing, software
//   distributed under the License is distributed on an "AS IS" BASIS,
//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
//   implied.  See the License for the specific language governing
//   permissions and limitations under the License.
																		
using System;
using System.Net;
using System.IO;
using System.Text;
using System.Collections.Generic;
using System.Linq;
using System.Json;

public class Naive
{
  // A useful function for encoding parameters into query strings
  static public string UrlEncode(Dictionary<string, string> dict)
  {
	return string.Join("&", dict.Select(
	  kvp => WebUtility.UrlEncode(kvp.Key) + "=" + WebUtility.UrlEncode(kvp.Value)
	).ToList());
  }
  
  // A useful function for joining paths into URLs
  static public string PathJoin(params string[] args)
  {
	return string.Join("/", args
					   .Select(arg => arg.Trim(new char[] {'/'}))
					   .Where(arg => !String.IsNullOrEmpty(arg))
	);
  }
  
  static public void Main ()
  {
	// Set everything up
	string key = "CHANGEME: YOUR_API_KEY";
	string secret = "CHANGEME: YOUR_API_SECRET";
	string version = "preview1";
	string practiceid = "000000";

	string baseUrl = "https://api.athenahealth.com/";

	// We use this a lot.
	Encoding UTF8 = System.Text.Encoding.GetEncoding("utf-8");
	
	// Easier to keep track of OAuth prefixes
	Dictionary<string, string> auth_prefixes = new Dictionary<string, string>()
	  {
		{"v1", "/oauth"},
		{"preview1", "/oauthpreview"},
		{"openpreview1", "/oauthopenpreview"},
	  };
	
	
	// Basic access authentication
	Dictionary<string, string> parameters = new Dictionary<string, string>()
	  {
		{"grant_type", "client_credentials"},
	  };
	
	// Create and set up a request
	WebRequest request = WebRequest.Create(PathJoin(baseUrl, auth_prefixes[version], "/token"));
	request.Method = "POST";
	request.ContentType = "application/x-www-form-urlencoded";
	
	// Make sure to add the Authorization header
	string auth = System.Convert.ToBase64String(UTF8.GetBytes(key + ":" + secret));
	request.Headers["Authorization"] = "Basic " + auth;

	// Encode the parameters, convert it to bytes (because that's how the streams want it)
	string encoded = UrlEncode(parameters);
	byte[] content = UTF8.GetBytes(encoded);

	// Write the parameters to the body
	Stream writer = request.GetRequestStream();
	writer.Write(content, 0, content.Length);
	writer.Close();
	
	// Get the response, read it out, and decode it
	WebResponse response = request.GetResponse();
	Stream receive = response.GetResponseStream();
	StreamReader reader = new StreamReader(receive, UTF8);
    JsonValue authorization = JsonValue.Parse(reader.ReadToEnd());
	
	// Make sure to grab the token!
	string token = authorization["access_token"];
	Console.WriteLine(token);
	
	// And always remember to close the readers and streams
	response.Close();
	
	
	// GET /departments
	
	// Since GET parameters go in the URL, we set up the parameters Dictionary first
	parameters = new Dictionary<string, string>()
	  {
		{"limit", "1"},
	  };
	
	// Now we get to make the URL, making sure to encode the parameters and remember the "?"
	string url = PathJoin(baseUrl, version, practiceid, "/departments", "?" + UrlEncode(parameters));

	// Create the request, add in the auth header
	request = WebRequest.Create(url);
	request.Method = "GET";
	request.Headers["Authorization"] = "Bearer " + token;
	
	// Get the response, read and decode
	response = request.GetResponse();
	receive = response.GetResponseStream();
	reader = new StreamReader(receive, UTF8);
	JsonValue departments = JsonValue.Parse(reader.ReadToEnd());
	Console.WriteLine(departments.ToString());

	response.Close();

	
	// POST /appointments/{appointmentid}/notes

	url = PathJoin(baseUrl, version, practiceid, "/appointments/1/notes");
	parameters = new Dictionary<string, string>()
	  {
		{"notetext", "Hello from C#"},
	  };

	request = WebRequest.Create(url);
	request.Method = "POST";
	request.ContentType = "application/x-www-form-urlencoded";
	request.Headers["Authorization"] = "Bearer " + token;

	content = UTF8.GetBytes(UrlEncode(parameters));
	writer = request.GetRequestStream();
	writer.Write(content, 0, content.Length);
	writer.Close();
	
	response = request.GetResponse();
	receive = response.GetResponseStream();
	reader = new StreamReader(receive, UTF8);
    JsonValue note = JsonValue.Parse(reader.ReadToEnd());
	Console.WriteLine(note.ToString());
	
	response.Close();
  }
}

File: samplecode/go/src/athenahealth.com/naive/naive.go
//    Copyright 2014 athenahealth, Inc.
//
//   Licensed under the Apache License, Version 2.0 (the "License"); you
//   may not use this file except in compliance with the License.  You
//   may obtain a copy of the License at
//
//       http://www.apache.org/licenses/LICENSE-2.0
//
//   Unless required by applicable law or agreed to in writing, software
//   distributed under the License is distributed on an "AS IS" BASIS,
//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
//   implied.  See the License for the specific language governing
//   permissions and limitations under the License.

package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
	"strings"
)

func main() {
	key := "CHANGEME: YOUR_API_KEY"
	secret := "CHANGEME: YOUR_API_SECRET"
	version := "preview1"
	practiceid := "000000"

	baseUrl := "https://api.athenahealth.com/"
	authPrefixes := map[string]string{
		"v1":           "/oauth",
		"preview1":     "/oauthpreview",
		"openpreview1": "/oauthopenpreview",
	}

	// client is used for all requests.
	client := &http.Client{}

	// Perform basic access authentication.  Make sure to make it a POST and to include grant_type
	// and Content-Type.
	authUrl := urlJoin(baseUrl, authPrefixes[version], "/token")
	parameters := url.Values{
		"grant_type": {"client_credentials"},
	}
	req, err := http.NewRequest("POST", authUrl, bytes.NewBufferString(parameters.Encode()))
	if err != nil {
		panic(err)
	}

	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
	req.SetBasicAuth(key, secret)

	// Send the request...
	resp, err := client.Do(req)
	if err != nil {
		panic(err)
	}
	// ... read the body of the response...
	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		panic(err)
	}
	// ... (make sure to close the body when you're done)...
	defer resp.Body.Close()

	// ... decode/unmarshal the JSON...
	var decoded map[string]interface{}
	err = json.Unmarshal(body, &decoded)
	if err != nil {
		panic(err)
	}

	// ... and store the token.
	token := decoded["access_token"]
	fmt.Println(token)

	parameters = url.Values{
		"limit": {"1"},
	}
	reqUrl := urlJoin(baseUrl, version, practiceid, "/departments")

	// GET requests take parameters in the URL, and they have to be encoded properly.
	req, err = http.NewRequest("GET", reqUrl+"?"+parameters.Encode(), nil)
	if err != nil {
		panic(err)
	}

	// Make sure to include the Authorization header and token in future requests.
	req.Header.Add("Authorization", fmt.Sprintf("Bearer %s", token))

	resp, err = client.Do(req)
	if err != nil {
		panic(err)
	}
	body, err = ioutil.ReadAll(resp.Body)
	if err != nil {
		panic(err)
	}
	defer resp.Body.Close()

	var departments map[string]interface{}
	err = json.Unmarshal(body, &departments)
	if err != nil {
		panic(err)
	}
	fmt.Println(departments)

	parameters = url.Values{
		"notetext": {"Hello from Go!"},
	}
	reqUrl = urlJoin(baseUrl, version, practiceid, "/appointments/1/notes")

	// POST requests take parameters in the body, so they need to be encoded and then converted to
	// bytes.
	req, err = http.NewRequest("POST", reqUrl, bytes.NewBufferString(parameters.Encode()))
	if err != nil {
		panic(err)
	}

	// POSTs also need the correct Content-Type.  And don't forget the access token!
	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Add("Authorization", fmt.Sprintf("Bearer %s", token))

	resp, err = client.Do(req)
	if err != nil {
		panic(err)
	}
	body, err = ioutil.ReadAll(resp.Body)
	defer resp.Body.Close()

	var note map[string]interface{}
	err = json.Unmarshal(body, &note)
	if err != nil {
		panic(err)
	}
	fmt.Println(note)
}

func urlJoin(args ...string) string {
	var trimmed []string
	for _, arg := range args {
		trimmed = append(trimmed, strings.Trim(arg, "/"))
	}
	return strings.Join(trimmed, "/")
}
